subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java

```diff
@@ -60,39 +68,57 @@ public void run() {
 
             private void doRun() {
                 while (true) {
+                    DatagramPacket packet;
                     long lockId;
                     try {
-                        lockId = communicator.receive();
+                        packet = communicator.receive();
+                        lockId = communicator.decodeLockId(packet);
                     } catch (GracefullyStoppedException e) {
                         return;
                     }
-                    lock.lock();
                     Runnable action;
-                    try {
+
+                    lock.lock();
+                    if (contendedActionsRunning.contains(lockId)) {
```

Comment:
I don't understand this. The action is running, so I set it to null and do nothing further down? What does that mean?

The code is very hard to follow. I suggest extracting a State class with all these maps and giving that class easy to understand methods.